zaza/openstack/utilities/openstack.py

```diff
@@ -1950,29 +2004,83 @@ def get_overcloud_auth(address=None, model_name=None):
-            os.makedirs(_dir)
-
-        model.scp_from_unit(
-            unit,
-            KEYSTONE_REMOTE_CACERT,
-            KEYSTONE_LOCAL_CACERT)
-
-        if os.path.exists(KEYSTONE_LOCAL_CACERT):
-            os.chmod(KEYSTONE_LOCAL_CACERT, 0o644)
-            auth_settings['OS_CACERT'] = KEYSTONE_LOCAL_CACERT
+    local_ca_cert = get_remote_ca_cert_file('keystone', model_name=model_name)
+    if local_ca_cert:
+        auth_settings['OS_CACERT'] = local_ca_cert
 
     return auth_settings
 
 
+async def _async_get_remote_ca_cert_file_candidates(application,
+                                                    model_name=None):
+    """Return a list of possible remote CA file names.
+
+    :param application: Name of application to examine.
+    :type application: str
+    :param model_name: Name of model to query.
+    :type model_name: str
+    :returns: List of paths to possible ca files.
+    :rtype: List[str]
+    """
+    cert_files = []
+    for _provider in CERT_PROVIDERS:
+        tls_rid = await model.async_get_relation_id(
+            application,
+            _provider,
+            model_name=model_name,
+            remote_interface_name='certificates')
+        if tls_rid:
+            cert_files.append(
+                REMOTE_CERT_DIR + '/' + CACERT_FILENAME_FORMAT.format(
+                    _provider))
+    cert_files.append(KEYSTONE_REMOTE_CACERT)
+    return cert_files
+
+_get_remote_ca_cert_file_candidates = zaza.model.sync_wrapper(
+    _async_get_remote_ca_cert_file_candidates)
+
+
+def get_remote_ca_cert_file(application, model_name=None):
+    """Collect CA certificate from application.
+
+    :param application: Name of application to collect file from.
+    :type application: str
+    :param model_name: Name of model to query.
+    :type model_name: str
+    :returns: Path to cafile
+    :rtype: str
+    """
+    unit = model.get_first_unit_name(application, model_name=model_name)
+    local_cert_file = None
+    cert_files = _get_remote_ca_cert_file_candidates(
+        application,
+        model_name=model_name)
+    for cert_file in cert_files:
+        _local_cert_file = "{}/{}".format(
+            LOCAL_CERT_DIR,
+            os.path.basename(cert_file))
+        with tempfile.NamedTemporaryFile(mode="w", delete=False) as _tmp_ca:
+            try:
+                model.scp_from_unit(
+                    unit,
+                    cert_file,
+                    _tmp_ca.name)
+            except JujuError:
+                continue
+            # ensure that the path to put the local cacert in actually exists.
+            # The assumption that 'tests/' exists for, say, mojo is false.
+            # Needed due to:
+            # commit: 537473ad3addeaa3d1e4e2d0fd556aeaa4018eb2
+            _dir = os.path.dirname(_local_cert_file)
+            if not os.path.exists(_dir):
+                os.makedirs(_dir)
+            shutil.move(_tmp_ca.name, _local_cert_file)
```

Comment:
I think the idea is that "move" is atomic, so your intention is good, but keep in mind that it's not atomic if done across two different filesystems, and I think here you're moving from a tmpfs to `/usr` so it won't be atomic

Reply:
In this case its the fact that shutil.move works across filesystem boundaries that made me opt for it rather than os.rename. Whether the operation is atomic or not it not too much of an issue here.